Hydro model channel improvements

matthewhoffman opened the pull request:
This pull request introduces a number of developments to get the subglacial hydrology model running reliably with channels active.

Most significant changes are:
* A bug fix to eliminate uphill outflow at terrestrial margins
* Disabling the calculation of iceThicknessHydro which was causing ice thickness artifacts near grounding lines
* Adding a namelist option for a maximum channel area to prevent runaway channel growth

Less important changes are:
* An option to disable terrestrial outflow entirely.  This has little impact on AIS simulations, but I left it in the code as an option.
* Some minor performance & timer adjustments
* Changes to disable channel melt inside lakes that were subsequently removed.  This introduced new problems so I removed the changes, but I left the commits in the history for posterity

matthewhoffman committed:
Add option to disallow terrestrial outflow

Desirable for AIS or for the situation where we do not consider
thermodynamics.

matthewhoffman committed:
Move assignment of melt input outside of SGH subcycling loop

When externally provided, these melt inputs will not change during the
SGH subcycling, so they don't need to be evaluated within the subcycling
loop.  This also moves one halo update out of the subcycle loop.

matthewhoffman committed:
Add more timers to hydrology solve

Useful for profiling where performance gains might occur

matthewhoffman committed:
Prevent uphill outflow at terrestrial margins

This commit fixes a bug that was introduced when downhill terrestrial
outflow was modified to ignore to the downhil bedtopo slope.  This
commit ensures that there is no outflow if the bedtopo external to the
grounded ice is a higher elevation than the bedtopo under the ice.  This
was previously occurring.

matthewhoffman committed:
disable channel melt into/out of lakes

matthewhoffman committed:
Allow channel melt out of lakes but not into or within

matthewhoffman committed:
Remove code that disables channel melt in lakes

This seemed to be a failed experiment as it prevents channels from ever
getting very extensive.  I'm keeping the previous commits in the history
for posterity.

matthewhoffman committed:
Disable iceThicknessHydro option by default

In AIS simulations, having iceThicknessHydro enabled was causing massive
lake formation and lack of channelization.  On further investigation, I
found there are thickened artifacts of increased thickness adjacent to
grounding lines.  This was in effect creating berms in the
hydropotential, making it harder for water to reach the ocean.  The
iceThicknessHydro calculations should be improved to eliminate this
issue, or, if that's not possible, removed from the code.  But for now I
am disabling it by default.

Note that I created an MPAS-Tools script to
preprocess ice thickness to eliminate depressions in the surface
elevation, and this greatly improves model stability around subglacial
lakes.  That preprocessing step should be used instead of
iceThicknessHydro.  It is unclear if that will be sufficient in the case
of evolving ice thickness.  If not, the surface-depression-filling
algorithm it employs should be ported to MALI and replace the existing
iceThicknessHydro algorithm.

matthewhoffman committed:
Add max channel area option

matthewhoffman commented:
Test run with input files is here on perlmutter: `/pscratch/sd/h/hoffman2/AIS_subglacial_hydro/dist_only/hager_high_res_params/output`
Note the input file had the ice thickness preprocessed to fill surface depressions using script in https://github.com/MPAS-Dev/MPAS-Tools/pull/520
Also note this run uses max channel area of 100km^2, but I subsequently set the default in the PR to 500 following https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2023JF007421.  The run uses the hydro params from the high-res run in Hager et al. 2022 (noting a typo in the kq value listed on p.3582 - I pulled the files from the archive to confirm)

Mass balance & time series from plotting script in https://github.com/MPAS-Dev/MPAS-Tools/pull/520 :
<img width="1798" height="694" alt="image" src="https://github.com/user-attachments/assets/c55b28ea-6af7-472e-885f-0855916a14aa" />

Comparison of water thickness within 1% of bump height (light shading) compared to radar specularity for Thwaites:
<img width="709" height="1145" alt="image (25)" src="https://github.com/user-attachments/assets/7925db1f-bf96-4ced-aaa9-90ef50d580ef" />

Comparison for more of WAIS (obs from https://royalsocietypublishing.org/doi/full/10.1098/rsta.2014.0297), lakes in peach, water thickness within 1% of bump height in light shading:
<img width="1373" height="754" alt="image" src="https://github.com/user-attachments/assets/5b3148a2-551e-4de6-97f3-adeb5a643bf0" />

alexolinhager commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -1994,6 +1998,16 @@ subroutine update_channel(block, err)
+            channelMelt(iEdge) = 0.0_RKIND
+            channelPressureFreeze(iEdge) = 0.0_RKIND
+         endif
+      enddo
```
We had previously tried something similar to this (https://github.com/E3SM-Project/E3SM/commit/c01e547d9a4de20ebf8cfbf4dafeabf696174dcc), but had done so by disabling all channel evolution instead of just disabling channelMelt and channelPressureFreeze over lakes. The other attempt never got merged because we still couldn't get stable channels, but this seems like a better approach, especially when paired with the other commits in this PR

alexolinhager replied:
Reading two more commits down the line, I think the issue with this approach is familiar and seems to be 
why we discontinued it previously

alexolinhager commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -78,8 +78,8 @@
 		
-		<nml_option name="config_SGH_use_iceThicknessHydro" type="logical" default_value=".true." units="unitless"
-			    description="Option to use an altered ice thickness field called iceThicknessHydro that replaces local maxima/minima in upperSurface with a mean of the cells neighbors. This option has no significant effect on the behavior of the model but makes it more stable."
+		<nml_option name="config_SGH_use_iceThicknessHydro" type="logical" default_value=".false." units="unitless"
```
I think we will eventually need to port your MPAS-Tools ice surface alteration code into MALI if we want to run with evolving dynamics. The primary reason for iceThicknessHydro was the development if single-cell surface depressions that formed when running coupled simulations. This was mostly an issue when running with the higher order velocity solver, which created speckled ice thicknesses along the domain edges as the run progressed.

It seems like the MPAS-Tools code should fix that issue, so I think we should try replacing the iceThicknessHydro code with what you wrote there. Would it be better to make that it's own PR, or tack it onto this one?

matthewhoffman replied:
Thanks for reminding me about the noise along the boundaries of regional domains with the higher-order advection.  I think change to iceThicknessHydro should occur in a separate PR.  The depression-filling algorithm in my standalone script will be a decent amount of work to port from python to fortran, I think, because it currently relies on some python libraries for a lot of the work.  Alternatively, we might be able to keep the existing iceThicknessHydro but disable modifications near the GL.  And maybe there is a fix to the HO advection scheme to ensure it reduces to first-order near the boundaries of regional domains (cc: @trhille).   With any of those options, I think we should tackle them separately from this PR.

alexolinhager commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2122,6 +2124,7 @@ subroutine evolve_channel(block, err)
       channelArea = channelChangeRate * deltatSGH + channelArea
       ! If sheet dissipation contributes to channel, there should be no need for a minimum channel size
       channelArea = max(1.0e-8_RKIND, channelArea)  ! make some tiny value when it goes negative
+      channelArea = min(config_SGH_chnl_max_area, channelArea) ! limit channel area to config-specified max value
```
Would it be better to just limit channelMelt and channelPressureFreeze when channelArea exceeds config_SGH_chnl_max_area? That way the channel still wouldn't grow past the max value but we also wouldn't violate mass conservation if there was no actual unaccounted melting. Probably doesn't affect the outcome but could be cleaner to not have a cell where channelMelt far exceeds the channelArea

matthewhoffman replied:
@alexolinhager , that's a good suggestion.  I'll look into how best to make the adjustment.

alexolinhager reviewed (commented):
Thanks for addressing these issues with the hydro model Matt! I made a few minor comments/suggestions, but otherwise should be ready to merge

matthewhoffman committed:
Change from max channel area to area where melt shuts off

This avoids the inconsistency of calculating melt but not letting it
change channel area.

matthewhoffman commented:
I reran the previous spinup with the new changes (disabling channel melting when channel area exceeds a max value rather than directly limiting channel area), plus increasing the channel area max value from 100 to 500 m^2.

The spinup still runs successfully.  Results are very similar, with the only notable difference in the global metrics being total lake volume increased from 5 Gt to 19 Gt.  I don't have a good explanation for that but it seems plausible these changes could lead to that (presumably the change in max channel area), so I'm comfortable to proceed with completing the review and merging.  @alexolinhager , let me know what you think.

<img width="1637" height="634" alt="image" src="https://github.com/user-attachments/assets/7ec837ac-0148-4731-b6a8-750837371ca4" />

matthewhoffman commented:
I force-pushed the last commit fixing a bug that was inadvertently disabling channel melting at **all** edges whenever a single edge exceeded the maximum channel area parameter.  After rerunning the multi-century spinup with the same parameters, I get a slightly different answer but qualitatively similar.  Channel outlux is now a slightly higher fraction of the total, which makes sense in that before (plot above) channel growth was potentially being limited.  Note that it does look very similar to the initial implementation where channel area was hardcapped at 100.
<img width="1794" height="693" alt="image" src="https://github.com/user-attachments/assets/37f375f4-c21f-4e1d-bbd0-4ccde21d9c45" />

Effective pressure and channel discharge for ASE also looks very similar to map shown above when channel area was hardcapped at 100.

<img width="1106" height="1085" alt="image" src="https://github.com/user-attachments/assets/b7b04d21-9e80-48e4-a767-5a971bb612e1" />

alexolinhager reviewed (approved):
@matthewhoffman, these changes look good to me. Just to make sure, I also tested this PR with the full integration suite, which passed the restart and decomposition tests. The hydro baseline comparisons failed, but that's to be expected with these changes
